### ISSUE TITLE
IGNORE_MOD_TAP_INTERRUPT support?

### ISSUE THREAD

**Issue by bomberstudios** [pqrs-org/Karabiner-Elements#2348]
Hi there!

I've recently added a complex modification to Karabiner, to turn the home row keys into modifiers. This is something I use in my external keyboard using QMK, and was really missing on my laptop keyboard. The modification looks like this:

```json
{
    "from": {
        "key_code": "f",
        "modifiers": {
            "optional": [
                "any"
            ]
        }
    },
    "to": [
        {
            "key_code": "left_shift"
        }
    ],
    "to_if_alone": [
        {
            "key_code": "f"
        }
    ],
    "type": "basic"
}
```

That's just an example, the full modification goes like this:
- A and ; are Control when held
- S and L are Alt when held
- D and K are Command when held
- F and K are Shift when held

This is a huge ergonomic improvement, and works reasonably well when typing slowly. However, if you type with any significant speed, you'll soon hit a problem: rolling pairs that involve the home row keys don't work at all (you'll get `ˆ` when you tried to type `li`, or `O` when you tried to type `fo`).

QMK has a setting (`IGNORE_MOD_TAP_INTERRUPT`) that solves this problem in an almost magical way: https://github.com/qmk/qmk_firmware/blob/master/docs/tap_hold.md#ignore-mod-tap-interrupt

I wonder if something similar could be added to Karabiner (or if it can be emulated with a combination of existing features).

Thanks in advance!

**Comment by moutis** [comment]
I tried this with Karabiner a year ago, but gave up because of all the ghosting. Now I'm moving into a QMK kbd and loving the return of home-row mods. I too would love to see Karabiner find a way to make this work. (I contribute to Karabiner dev almost once/year...thanks for this great tool)

**Comment by Rougemoot** [comment]
Hello, just commenting on this because this is something I would also like to see happen in karabiner. Some of QMK's options are sorely missed, and this is one of the more popular ones I think.

**Comment by MuhammedZakir** [comment]
IIRC, using `from.simultaneous`[1] solves this problem. You might need to set `basic.simultaneous_threshold_milliseconds` to a low value.

[1] https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/from/simultaneous/

**Comment by Rougemoot** [comment]
I've tried many things and I can't get the desired result. How would I set the simultaneous stuff to get working home row mods?

**Comment by MuhammedZakir** [comment]
> I've tried many things and I can't get the desired result. How would I set the simultaneous stuff to get working home row mods?

Try below config. Depending on your typing speed, you may need to set `basic.simultaneous_threshold_milliseconds` to a low value. And depending on usage, you may also need to set `simultaneous_options.key_down_order` to `strict` [1].

[1] https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/from/simultaneous-options/#key_down_order

```json
    {
      "description": "a & s => shift",
      "manipulators": [
        {
          "type":"basic",
          "from":{
            "modifiers":{
              "optional":[
                "any"
              ]
            },
            "simultaneous":[
              {
                "key_code":"a"
              },
              {
                "key_code":"s"
              }
            ],
            "simultaneous_options":{
              "key_up_when":"all"
            }
          },
          "to":[
            {
              "key_code":"shift"
            }
          ]
        }
      ]
    }
```

**Comment by Rougemoot** [comment]
Does it have to be a combo? I'd just want `a` to be shift, not the `as` combo.

**Comment by MuhammedZakir** [comment]
> Does it have to be a combo? I'd just want `a` to be shift, not the `as` combo.

How would you type `a` then? Can you please explain when do you want `a` to be `a` and when do you want `a` to be `shift`?

**Comment by Rougemoot** [comment]
The way I have it on my mechanical keyboard with QMK is this:

`A` is `a` when typed pressed alone and released withing 200ms.  
`A` is `shift` when pressed with another key *that is released before `A`*, even before the 200ms are up.  
`A` is `shift` when pressed alone and released *after* 200ms. This helps when you think you want to use shift, and then change your mind; releaseing the key, even by itself, doesn't post an `a`.  
`A` is `shift` when pressed with another key, even in that key is released after `A`, *if the 200ms were up by the time the second key was released*.

So to sum up, if I press `A`, it's gonna post `a` if released by itself, `shift` if the 200ms are up when releasing it, *unless* the second key is released before `A`, regardless of timing.

**Comment by MuhammedZakir** [comment]
You can use `to_if_alone`[1] to achieve what you want. And set `basic.to_if_alone_timeout_milliseconds` to `200` and it will work (I think). See the example in doc and "About cancellation" section below that.

[1] https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/to-if-alone/

**Comment by Rougemoot** [comment]
The problem with `to_if_alone` is *exactly* that we don't have an `IGNORE_MOD_TAP_INTERRUPT`-like option (circling back to the whole point of this issue).

From "About cancellation":

> to_if_alone is canceled if other events (keys, buttons or scroll wheel) is happen while the from key is pressed down.

What `IGNORE_MOD_TAP_INTERRUPT` does is ignore that cancellation if the second keypress isn't released before the `from` event, solving the issues with overlapping keypresses when typing fast.

With `if_alone`, if you do

```
A keyDown
I keyDown
A keyUp
I keyUp
```

before the timeout is up, you get `I` as an output, not `ai`, which is bad, because overlapping keypresses are incredibly common when typing fast.  
If there was such a thing as `IGNORE_MOD_TAP_INTERRUPT`, the output would be `ai`, as intended, and you'd have to be careful to release `I` before `A` to get a capitlized `I` output, or wait for the timeout to be elapsed before you press the `I` key (at which point release order doesn't matter).

This is very important if you want home row mods such as "Shift on A", otherwise you get false positives all over the place and the whole thing becomes unusable.

**Comment by MuhammedZakir** [comment]
> The problem with `to_if_alone` is _exactly_ that we don't have an `IGNORE_MOD_TAP_INTERRUPT`-like option (circling back to the whole point of this issue).
> _--snip--_

Oops! :grimacing: Then, I guess `from.simultaneous` is the only option atm? However, you will need two keys for that.

**Comment by Rougemoot** [comment]
Yes, I guess so. This is a feature I think many people would like to see make its way into Karabiner. Fingers crossed I guess?

**Comment by Rougemoot** [comment]
I think one way to implement this is to add an option which, when set, makes `if_alone` get interrupted but a key **up** event rather than a key down event. I'm not sure how karabiner works behind the scenes but I guess this might make sense?

**Comment by gilescope** [comment]
Home row mods are the future!

**Comment by franco** [comment]
@tekezo are there any plans regarding this?

**Comment by franco** [comment]
It looks like [kmonad](https://github.com/david-janssen/kmonad) is a good alternative which supports [row mod](https://precondition.github.io/home-row-mods).

**Comment by Rougemoot** [comment]
Yes, but kmonad doesn't (yet) have support for combos (like to.simultaneous), and is to this day still a bit janky on macOS.

Couldn't this be done in Karabiner just by adding an option to look for key *releases* rather than keypresses in the `to_if_alone` logic?

**Comment by ryanjanssen** [comment]
+1 for this question - I would LOVE mod_tap_interrupt in Karabiner.  It works wonders on my mechanical keyboard

**Comment by fabiomcosta** [comment]
This is driving me crazy as well :/
I tried literally everything, but there is no way to make it work properly today without Karabiner changes :/
I wish I was able to contribute a fix for this easily :/

**Comment by orand** [comment]
+1 This is important to me as well.

**Comment by notnotjake** [comment]
I would really love to have Home Row Mods on my laptop keyboard after getting used to them on a QMK keyboard. Is there any way to try to get this added to the roadmap?

**Comment by ajit283** [comment]
Hi, I had this exact problem (I think) and I cobbled something together after not finding anything adequate: https://github.com/ajit283/keyboardremapper/tree/main
It's still very basic, as I just need to mod-tap the spacebar and IJKL arrows, but one could maybe make it more extensible!

**Comment by gabriel-gardner** [comment]
I don't think I've figured out how to emulate all the behavior @Vermoot asks for, but I have been able to get Karabiner-Elements to emulate simple modtap functionality by setting up a goku config like in this contrived example: 

https://github.com/gabriel-gardner/modtap-karabiner/blob/main/modtap_example.edn

With this setup you can have a key that if held longer than `to_if_alone_timeout_milliseconds` it becomes a modifier, but if interrupted before just outputs both keys like they were typed regularly. Held over the tapping term, the key acts like a modifier on any other keys that you setup rules for, and gets canceled (outputting the canceled key) if an interaction isn't defined. I found this far more accurate than trying to use `from.simultaneous` as it would accidentally trigger when typing quickly.

I don't use this for modifiers like turning `a` to `Shift` (I use QMK for that), but I think it can be setup that way, if you define all the rules for the `a` modifier to output the appropriate capitalized letter. I personally use it to map key combinations to macros (ie holding `o` over tapping term then typing `c` will open Chrome, or typing `s` will open Safari, etc). Perhaps this could be helpful for some people here!

**Comment by ldebritto** [comment]
Using a lot of simlayers and possibly a bunch of unneeded code, I think I may have recreated the desired behavior. At least, that's what I'm rocking for a while without any major issues.

I've isolated the home row mods to one file for easier reading should anyone be interested in copying it:

https://github.com/ldebritto/karabiner-config/blob/main/karabiner%20-%20home%20row%20mods.edn

**Comment by eugenesvk** [comment]
based on @gabriel-gardner's bright idea I've created [this Goku gist](https://gist.github.com/eugenesvk/1f8fee49ce3fa5fe3cd4cce52085f7a5) with home row mods (currently with  limited 2-key combo support among ⌥⌘⇧, and ^⇧; see limitations at the top of the gist)
Would be great if you could try it out to see whether this is indeed the better solution in the absence of the proper QMK-like experience and what timings would work best with it.
To avoid messing up with the global timings used in other combos might be best to
- set timing (via search&replace as the timings are set in each of the mode layers and are the same) to unique numbers not present in the main profile (from the comment above `:alone 194` should be smaller than the `:held 195 :delay 195`)
- install this to a `goku` profile
- copy&paste the `goku` profile to the main profile
- easily globally search&replace the timings only within home row mods (since they are different from all other profile timings)

If this works more or less fine then it would still need a feature to be able to set multiple variables, which might also need a change in Karabiner :(, but it might be an easier one than this one

**Comment by adimassalleh** [comment]
Hi All! This is my first post in Github.

After working on it for the last two days, I think I've cracked the problem. The approach used is to halt the invocation of key up which is required if the press is incomplete cycle such when another key interferes before the main key got released. This is an example.

      "from": {
        "key_code": "a"
      },
      "to_if_held_down": [{
        "key_code": "left_control",
	"lazy": true,
	"halt": true
      }],
      "to_if_alone": [{
        "key_code": "a",
	"halt": true
      }],
      "to_after_key_up": [{
        "key_code": "a"
      }]

There is still room for improvements but at least it is workable. Some issue still remains such as typing the word 'as' quickly or any quick consecutive modifier keys which might not register the full keys completely.

**Comment by gabriel-gardner** [comment]
@jeremiahrose glad that setup works for you. If you would like a solution that avoids having to tweak delays, however, the config I linked above allows you to wait indefinitely to type `h` after pressing `f` to get `H`. Instead, to type `f` you just type it faster than the modtap delay, just like how it is handled in qmk. I've been using a configured version of the config I linked above for more than a year and it works flawlessly to get homerow mod behavior, with minimal code as well. Great to see people are still looking into this and hopefully we can eventually get default support from Karabiner!

**Comment by willpuckett** [comment]
It doesn't seem like this issue should be closed. Direct home row mods are not working; all the suggestions require a mod key to function reliably with the currently available configuration options. Using solely the homerow keys directly as modifiers requires modifying typing style to be very mechanical, in particular, avoiding rolls. 

In an ideal world, it would be possible to configure them to work like the following zmk config ([documentation](https://zmk.dev/docs/behaviors/hold-tap#positional-hold-tap-and-hold-trigger-key-positions):
```
    hrml: home_row_mod_left {
        compatible = "zmk,behavior-hold-tap";
        label = "HOME_ROW_MOD_LEFT";
        #binding-cells = <2>;
        flavor = "balanced";
        tapping-term-ms = <HM_TAPPING_TERM>;
        hold-trigger-key-positions = < KEYS_R THUMBS >;
        hold-trigger-on-release;
        bindings = <&kp>, <&kp>;
        };
```

 Yes, it's a feature request! Having parity between mechanical keyboard configs and the built in keyboard is a major usability issue to ease switching back and forth. It also opens up the command keys to be used for backspace/enter which is much more comfortable for users with shorter pinkies.

**Comment by eugenesvk** [comment]
> Using solely the homerow keys directly as modifiers requires modifying typing style to be very mechanical, in particular, avoiding rolls

That doesn't seem to be the case, the method linked above does not require any rolls changes, if 
- <kbd>f</kbd> is <kbd>⇧</kbd> key and 
- <kbd>d</kbd> is <kbd>⌘</kbd>


then rolling taps: down<kbd>f</kbd>down<kbd>d</kbd>up<kbd>f</kbd>up<kbd>d</kbd> types `fd` right away. Now, if <kbd>d</kbd> is held down longer, you'll only get an `f`, and a held <kbd>⌘</kbd>, but that's expected.

This still doesn't make the home row mods perfect as they're timing-based, so would indeed be very nice if Karabiner-Elements got a reliable fully functional home row mod functionality (e.g., if you continue to hold <kbd>f</kbd> but type <kbd>d</kbd> before a delay is up, you'll still get `fd`)

**Comment by jaroslaw-weber** [comment]
agree, would be nice to have 1 setting instead of thousands lines of json that's impossible to decode. can we reopen this issue?

**Comment by willpuckett** [comment]
@jaroslaw-weber I ultimately ended up using [karabiner.ts](https://github.com/evan-liu/karabiner.ts) to generate the config file. While the home row mods still don't feel as comfortable as my mechanical keyboard, I was able to use the js spread operator to drastically reduce the config to a much more readable and maintainable volume. An example using the Engram layout is [here](https://github.com/willpuckett/karabiner-config-engram/blob/main/index.ts).

It seems like this issue could also be filed with Apple. I'm filing it as a bug via the [Feedback Assistant App](https://developer.apple.com/bug-reporting/) and would encourage others to do so as well. While it is wonderful to be able to modify keyboard behavior via Karabiner, implementing some of these deeper behaviors might be more appropriate at a deeper level. There is so much excellent work going on in the keyboarding community and I believe it is a fair expectation that some of the successes begin to make their way into commercial products.

**Comment by cchaglund** [comment]
> karabiner.ts

Based on the upvotes it seems like you've managed to figure some stuff out! Could you perhaps explain a bit more about what you've done, especially for those of use that aren't used to the edn format?

**Comment by willpuckett** [comment]
@amiorin got this one. Just used his guide https://gist.github.com/amiorin/4c74f63fe599a1dcbd0933628df1aac9 to switch to Kmonad. It's much smoother and way more concise.

**Comment by ayyjohn** [comment]
> Using a lot of simlayers and possibly a bunch of unneeded code, I think I may have recreated the desired behavior. At least, that's what I'm rocking for a while without any major issues.
> 
> I've isolated the home row mods to one file for easier reading should anyone be interested in copying it:
> 
> https://github.com/ldebritto/karabiner-config/blob/main/karabiner%20-%20home%20row%20mods.edn

this worked wonders for me! I've got a keyboard with ZMK and i've been searching for something close enough for my macbook. I'd been running into a ton of unfortunate crashes with kmonad :( 

if you want to modify it to have a different arrangement and you're new to goku, make sure to edit all the values all the way down (find and replace should get you most of the way). at first i didn't notice what was happening and couldn't figure out why the changes weren't working, but i'd only changed the home row values.
